python/cuml/dask/common/base.py

```diff
@@ -0,0 +1,224 @@
+        tasks that can execute concurrently on each worker.
+
+        Note that this mixin assumes the subclass has been trained and
+        includes a `self.local_model` attribute containing a subclass
+        of `cuml.Base`.
+
+        This is intended to abstract functions like predict, transform, and
+        score, which can execute embarrassingly parallel but need addition
+        execution constraints which result from the more limited GPU
+        resources.
+
+        Parameters
+        ----------
+        func : dask.delayed function to propagate to the workers to execute
+               embarrassingly parallel, shared between tasks on each worker
+               and constrained by a holder lock
+
+        X : Dask cuDF dataframe  or CuPy backed Dask Array (n_rows, n_features)
+            Distributed dense matrix (floats or doubles) of shape
+            (n_samples, n_features).
+
+        delayed : bool return a lazy (delayed) object?
+
+        parallelism : int
+            Amount of concurrent partitions that will be processed
+            per worker. This bounds the total amount of temporary
+            workspace memory on the GPU that will need to be allocated
+            at any time.
+
+        output_futures : bool returns the futures pointing the to the resuls
+                         of the parallel function executions on the workers,
+                         rather than a dask collection object.
+
+        Returns
+        -------
+        y : dask cuDF (n_rows, 1)
+        """
+
+        if delayed:
+            X_d = X.to_delayed()
+
+            lock = dask.delayed(MultiHolderLock(parallelism),
+                                pure=True)
+
+            model = dask.delayed(self.local_model, pure=True)
+
+            func = dask.delayed(func, pure=False, nout=1)
+
+            if isinstance(X, dcDataFrame):
+                preds = [func(model, lock, part) for part in X_d]
+                return preds if output_futures \
+                    else dask.dataframe.from_delayed(preds)
+
+            else:
+                preds = [func(model, lock, part[0])
+                         for part in X_d]
+
+                # todo: add parameter for option of not checking directly
+                dtype = X.dtype
+
+                preds_arr = [
+                    dask.array.from_delayed(pred,
+                                            meta=cp.zeros(1, dtype=dtype),
+                                            shape=(np.nan,),
+                                            dtype=dtype)
+                    for pred in preds]
+
+                if output_futures:
+                    return preds_arr
+                else:
+                    return dask.array.concatenate(preds_arr, axis=0,
+                                                  allow_unknown_chunksizes=True
+                                                  )
+
+        else:
+            X = X.persist()
```

Comment:
The more I thought about this change, I now remember that we had originally designed the eager case so that it could continue to make use of, and apply an upper-bound to, worker-level parallelism by allowing each worker to use multiple streams to run a predict, transform, or inverse_transform function in parallel, even when `n_threads` has been set to 1. Unfortunately, grouping the partitions to do this worker-level parallelism is non-trivial to do in the lazy environment because we don't yet know which workers each partition is on. 

I think we should consider keeping the eager path as an option. I see two different cases where users might want to consider using lazy vs eager:
1) Exploratory data analysis user just wants to do some etl, train up a bunch of models, and will probably have to set `n_threads=1` because of the potential for deadlocks in NCCL during the synchronous training. Using the eager path still allows the user to benefit from parallellism during prediction.
2) The FIL-style user who just wants to load a trained model and go to town with predictions, possibly in production. They've likely started their cluster with a higher value for `n_threads` and can make use of lazy predictions, thus Dask will naturally schedule more concurrent tasks per worker and our current lazy approach will guarantee the model instance is not duplicated for each task. 

For both of these cases, I think we can remove the lock, unless it turns out to be a problem. I think the max_parallelism can go away as well because we are assuming the user has set that indirectly through Dask. The `n_internal_streams` (soon to be called `n_worker_streams`) can be used in the eager path through predict so they can be used to parallelize the partitions on each worker. 

What do you guys think?

Reply:
I definitely agree with point number 1, getting back the eager execution indeed.

Point 2 I agree but somwheat less so, as in I think for the linear models it will probably not be an issue to remove the lock, but I would like to see that change tested with FIL before 0.13 is done, in order to have more evidence that indeed it won't be a problem. So removing it from OLS/PCA/et.al and then testing on the RF PR, what do you think?

Reply:
> but I would like to see that change tested with FIL before 0.13 is done, in order to have more evidence that indeed it won't be a problem

From a technical standpoint, this lock is not going to add anything at all, other than allowing the user to further limit the number of concurrently executing tasks for different DAGS. The `n_threads` (https://distributed.dask.org/en/latest/worker.html#thread-pool) option in Dask already accomplishes this by limiting the number of concurrently executing tasks across all DAGs. 

The reason I used the statement "cause a problem" in my previous comment is because I don't know if users are actually  going to want or use that much fine-grained control. I'm just concerned now that keeping the lock in the fully lazy path is going to make it even more confusing for them to configure properly. I really could go either way on whether it should be included but i want to be careful to understand that this lock is no longer serving a function purpose to memory management outside of convenience for users who want to increase n_threads to a significant number (like the number of cores for a 40-core machine). `LocalCUDACluster` already sets this value to 1 by default. That being said, the more I've thought about it today, the more I've been finding reasons to talk myself out of it. 

John is right, though- if we do decide to keep the lock in for now, we need to insert a sleep for a random amount of time or use an exponential backoff to limit the contention and crowding as `n_threads` becomes >> 1. I included this as a crude implementation just to demonstrate it's possible and I planned to include more tests in the PCA/TSVD PR. I want to be careful, though, and make sure not to overpromise the contents of that PR for this release, because there's a lot of other things going on for 0.13 currently.

Reply:
After all this, Dante and I talked a little bit more offline and I’ve swayed back in the direction of keeping the lock until we determine that it’s no longer useful. Over time if it becomes a burden to maintain or we determine it is more confusing than helpful, we can remove it.